File: pkg/ot/selection.go
package ot

type Selection struct {
	Anchor int
	Head   int
}

func TransformSelection(selection Selection, operation *Operation) Selection {
	transformIndex := func(index int) int {
		newIndex := index
		for _, op := range operation.Ops {
			switch v := op.(type) {
			case RetainOp:
				index -= v.N
			case InsertOp:
				newIndex += len(v.Text)
			case DeleteOp:
				newIndex -= min(index, v.N)
				index -= v.N
			}
			if index < 0 {
				break
			}
		}
		return newIndex
	}

	newAnchor := transformIndex(selection.Anchor)
	if selection.Anchor == selection.Head {
		return Selection{Anchor: newAnchor, Head: newAnchor}
	}
	return Selection{Anchor: newAnchor, Head: transformIndex(selection.Head)}
}

File: assets/src/selection.ts
import { isInsert, isRetain, Operation } from "./ot";

export type Selection = {
  anchor: number;
  head: number;
};

export function transformSelection(
  selection: Selection,
  operation: Operation,
): Selection {
  function transformIndex(index: number): number {
    let newIndex = index;
    const ops = operation.ops;
    for (let i = 0; i < ops.length; i++) {
      const op = ops[i];
      if (isRetain(op)) {
        index -= op.n;
      } else if (isInsert(op)) {
        newIndex += op.text.length;
      } else {
        newIndex -= Math.min(index, op.n);
        index -= op.n;
      }
      if (index < 0) break;
    }
    return newIndex;
  }

  const newAnchor = transformIndex(selection.anchor);
  if (selection.anchor == selection.head) {
    return { anchor: newAnchor, head: newAnchor };
  }
  return { anchor: newAnchor, head: transformIndex(selection.head) };
}
